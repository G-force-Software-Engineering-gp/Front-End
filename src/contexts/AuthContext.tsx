import { createContext, useState, useEffect } from "react";
import jwt_decode from "jwt-decode";
import { useNavigate } from "react-router-dom";
import axios from "axios";
// import swal from "sweetalert";
// import axiosInstance from "../utils/axiosInstance";


export type AuthContextType = {
  user: any; // Replace with your user type
  setUser: React.Dispatch<React.SetStateAction<any>>;
  authTokens: any; // Replace with your auth tokens type
  setAuthTokens: React.Dispatch<React.SetStateAction<any>>;
  loginUser: (values: any) => Promise<void>;
  registerUser: (values: any) => Promise<void>;
  logoutUser: () => void;
};

const AuthContext = createContext<AuthContextType | null>(null);

export default AuthContext;
type AuthProviderProps = {
    children: React.ReactNode;
  };
  
  export const AuthProvider: React.FC<AuthProviderProps> = ({ children }) => {
  const [authTokens, setAuthTokens] = useState<any>(
    localStorage.getItem("authTokens")
      ? JSON.parse(localStorage.getItem("authTokens")!)
      : null
  );
  const [user, setUser] = useState<any>(
    localStorage.getItem("authTokens")
      ? jwt_decode(localStorage.getItem("authTokens")!)
      : null
  );
  const [loading, setloading] = useState(true);

  const Navigate = useNavigate();

  const loginUser = async (values: any) => {
    const response = await fetch(
        "https://amirmohammadkomijani.pythonanywhere.com/auth/jwt/create/",
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            email: values.email,
            password: values.password,
          }),
        }
      );
      const data = await response.json();
      console.log(data)
      if (response.status === 200) {
        setAuthTokens(data);
        setUser(jwt_decode(data.access));
        localStorage.setItem("authTokens", JSON.stringify(data));
        Navigate("/");
      } 
      else{
        console.log(data.error);
      }
  };
  //axios approach
        // await axios.post("https://amirmohammadkomijani.pythonanywhere.com/auth/jwt/create/",
        //     {
        //     email: values.email,
        //     password: values.password,
        //     },
        //     {
        //     headers: {
        //     'Content-Type': 'application/json',
        //     },
        //     })
        //     .then((response) => {
        //     const data = response.data;
        //     console.log(data);
        
        //     if (response.status === 200) {
        //         setAuthTokens(data);
        //         setUser(jwt_decode(data.access));
        //         localStorage.setItem('authTokens', JSON.stringify(data));
        //         Navigate('/');
        //     } else {
        //         console.log(data.error);
        //     }
        //     })
        //     .catch((error) => {
        //     console.error('API request failed:', error);
        //     });
        // }
  const registerUser = async (values: any) => {
      const reg_response = await fetch(
      "https://amirmohammadkomijani.pythonanywhere.com/auth/users/",
      {
          method: "POST",
          headers: {
          "Content-Type": "application/json",
          },
          body: JSON.stringify({
          first_name: values.firstName,
          last_name: values.lastName,
          email: values.email,
          username: values.username,
          password: values.password,
          }),
      });
      console.log(reg_response);
      const reg_data = await reg_response.json();
      if (reg_response.status === 201) {
        Navigate("/login");
      } else {
          console.log(reg_data.error);  
      }
  };

  //axios approach
        // await axios.post("https://amirmohammadkomijani.pythonanywhere.com/auth/users/",
        //     {
        //     first_name: values.firstName,
        //     last_name: values.lastName,
        //     email: values.email,
        //     username: values.username,
        //     password: values.password,
        //     },
        //     {
        //     headers: {
        //     'Content-Type': 'application/json',
        //     },
        //     })
        //     .then((reg_data) => {
        //         const data = reg_data.data;
        //         console.log(data);
        //         if (reg_data.status === 201) {
        //           Navigate('/login');
        //         } else {
        //           console.log(data.error);
        //         }
        //       })
        //       .catch((error) => {
        //         console.error('API request failed:', error);
        //       });
        // }

  const logoutUser = () => {
    setAuthTokens(null);
    setUser(null);
    localStorage.removeItem("authTokens");
    Navigate("/login");
  };

  const contextData: AuthContextType = {
    user,
    setUser,
    authTokens,
    setAuthTokens,
    loginUser,
    registerUser,
    logoutUser,
  };

  useEffect(() => {
    if (authTokens) {
        setUser(jwt_decode(authTokens.access));
      }
      setloading(false);
  }, [authTokens, loading]);

  return <AuthContext.Provider value={contextData}>{children}</AuthContext.Provider>;
};
