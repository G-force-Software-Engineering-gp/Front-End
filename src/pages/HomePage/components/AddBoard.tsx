import { Button } from '@/components/ui/button';
import { Card } from '@/components/ui/card';
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogFooter,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from '@/components/ui/dialog';
import { Input } from '@/components/ui/input';
import { Label } from '@/components/ui/label';
import AuthContext from '@/contexts/AuthContext';
import { DialogClose } from '@radix-ui/react-dialog';
import React, { useContext, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import Swal from 'sweetalert2';

interface AddBoardProps {
  workspaceId: number; // Define the type for the workspaceId prop
}

const AddBoard: React.FC<AddBoardProps> = ({ workspaceId }) => {
  //   console.log(workspaceId);
  const [name, setname] = useState('');
  const [nameError, setnameError] = useState(false);
  let authTokens = useContext(AuthContext)?.authTokens;
  const navigate = useNavigate();
  const CreateWorkspace = async () => {
    const data = await fetch('https://amirmohammadkomijani.pythonanywhere.com/tascrum/crboard/', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `JWT ${authTokens.access}`,
      },
      body: JSON.stringify({
        title: name,
        workspace: workspaceId,
      }),
    }).then((response) => response);
    console.log(data);
    if (data.ok) {
      Swal.fire({
        icon: 'success',
        title: 'Successful',
        text: 'Workspace created successfully',
        timer: 3000,
      });
      setname('');
      setInterval(() => {
        navigate(0);
      }, 2000);
    } else {
      Swal.fire({
        icon: 'error',
        title: 'Error',
        text: 'Something went wrong!',
        timer: 3000,
      });
      setname('');
      setInterval(() => {
        navigate(0);
      }, 2000);
    }
  };

  const submitFn = () => {
    if (name.length === 0) {
      setnameError(true);
    } else {
      CreateWorkspace();
    }
  };

  return (
    <div>
      <Dialog>
        <DialogTrigger asChild>
          <Card className=" cursor-pointer bg-slate-300 dark:bg-slate-800">
            <div className="grid cursor-pointer place-items-center py-14 md:py-14" data-testId="clicking">
              <h1 className="text-xl font-semibold">Create a new Board</h1>
            </div>
          </Card>
        </DialogTrigger>
        <DialogContent className="sm:max-w-[425px]">
          <DialogHeader>
            <DialogTitle className="add-board-title">Add Board</DialogTitle>
            <DialogDescription>Enter name of the board</DialogDescription>
          </DialogHeader>
          <div className="grid gap-4 py-4">
            <div className="grid grid-cols-4 items-baseline gap-4">
              <Label htmlFor="name" className="text-right">
                Name
              </Label>
              <Input
                id="name"
                placeholder="Write name of your workspace"
                className={`col-span-3 ${nameError && name.length === 0 ? 'border-red-500' : ''}`}
                onChange={(e) => setname(e.target.value)}
              />
            </div>
          </div>
          <DialogFooter>
            <DialogClose>
              <Button className="add-board-button" onClick={() => submitFn()}>
                Add Board
              </Button>
            </DialogClose>
          </DialogFooter>
        </DialogContent>
      </Dialog>
    </div>
  );
};

export default AddBoard;
