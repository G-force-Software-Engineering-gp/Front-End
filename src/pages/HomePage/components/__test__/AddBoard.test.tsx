import { AuthProvider } from '@/contexts/AuthContext';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from 'react-router-dom';
import AddBoard from '../AddBoard';

describe('AddBoard Component', () => {
  test('renders AddBoard component', () => {
    render(
      <MemoryRouter>
        <AuthProvider>
          <AddBoard workspaceId={1} />
        </AuthProvider>
      </MemoryRouter>
    );
  });
  test('matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <AuthProvider>
          <AddBoard workspaceId={1} />
        </AuthProvider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('submits form with valid input', () => {
    render(
      <MemoryRouter>
        <AuthProvider>
          <AddBoard workspaceId={1} />
        </AuthProvider>
      </MemoryRouter>
    );
    const button = screen.getByTestId('clicking');
    fireEvent.click(button);
    const input = screen.getByLabelText('Name');
    fireEvent.change(input, { target: { value: 'Test Board' } });
    expect(input).toHaveValue('Test Board');
    // fireEvent.click(screen.getByText('Add Board', { selector: '.add-board-button' }));
  });
  test('renders Add Board button', () => {
    render(
      <MemoryRouter>
        <AuthProvider>
          <AddBoard workspaceId={1} />
        </AuthProvider>
      </MemoryRouter>
    );
    const button = screen.getByTestId('clicking');
    fireEvent.click(button);
    const addButton = screen.getByText('Add Board', { selector: '.add-board-button' });
    expect(addButton).toBeInTheDocument();
  });
  test('renders correct title and heading', () => {
    render(
      <MemoryRouter>
        <AuthProvider>
          <AddBoard workspaceId={1} />
        </AuthProvider>
      </MemoryRouter>
    );
    const button = screen.getByTestId('clicking');
    fireEvent.click(button);
    const title = screen.getByText('Add Board', { selector: '.add-board-title' });
    const heading = screen.getByText('Create a new Board');
    expect(title).toBeInTheDocument();
    expect(heading).toBeInTheDocument();
  });
});
