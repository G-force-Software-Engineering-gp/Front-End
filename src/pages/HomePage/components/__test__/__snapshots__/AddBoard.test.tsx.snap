// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBoard Component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-controls="radix-:r3:"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="rounded-lg border text-card-foreground shadow-sm cursor-pointer bg-slate-300 dark:bg-slate-800"
      data-state="closed"
      type="button"
    >
      <div
        class="grid cursor-pointer place-items-center py-14 md:py-14"
        data-testid="clicking"
      >
        <h1
          class="text-xl font-semibold"
        >
          Create a new Board
        </h1>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
